### ISSUE TITLE
Filter Widget New Search

### ISSUE THREAD

**Issue by doveyg** [Mottie/tablesorter#1070]
When I use the Filter Widget using the new method

$.tablesorter.setFilters($("#table"), array, true);

i get an error on line 1474 in the file jquery.tablesorter.widgets.js because .toLowerCase is not a function using Firefox.

Can this be replicated?

**Comment by Mottie** [comment]
Hi @doveyg!

Are you using the latest version? What does that line of code look like?

In the most recent version, I see this line:

```js
data.iFilter = wo.filter_ignoreCase ? ( data.filter || '' ).toLowerCase() : data.filter;
```

And the only way that would cause an error is if the contents of the array contain non-string values, like a number. If you want to search for a number, make it into a string, e.g. `10` becomes `"10"`.

I'll fix that line to prevent errors, but use the above method in the mean time.

**Comment by doveyg** [comment]
its the same line. 2.24.2?

**Comment by Mottie** [comment]
This is now fixed in the master branch. All filters in the array are automatically converted into strings.

**Comment by doveyg** [comment]
I'm not sure if this is the same issue, but; its now converting

[undefined, undefined, undefined, "/^Jessie$/i"]

to

[",,,/^Jessie$/i"]

as shown in firefox's console. previously, passing an array like this worked.

Net result, still broken.

**Comment by Mottie** [comment]
Hmm, what version of Firefox are you using? I just tried a similar search and it converted the undefined values into an empty string, not one big glob. Did you mention before that you are using Linux? I wonder if that might be the issue.

**Comment by doveyg** [comment]
I'm using windows. I will have to get back to you on the first part.

**Comment by Mottie** [comment]
Ok, I changed the method used to convert the filters into an array of strings in https://github.com/Mottie/tablesorter/commit/ac5818294dac69cbda714170961ca6b8e7d076a3. See if that works better for you.

**Comment by doveyg** [comment]
I lost the PC in a HDD failure. I am now using the latest version of Chrome, It is converting properly. But its still not filtering.

here is the code;

    var options = {"sorttable":"#batch-sort","sortorder":[[2,0]],"rowheaders":"batch","headers":{"1":{"sorter":"duration"},"4":{"columnSelector":false}}}

    $(options.sorttable).tablesorter({
    cssChildRow: "hiddenrow",
	    widthFixed:true,
		sortList: options.sortorder, //[]
		headers: options.headers, //{}
		//textAttribute: "data-duration",
		widgets: ["zebra", "filter", "columnSelector", "pager", "output"], //math
		widgetOptions : {
			output_delivery: "download",
			output_saveRows: $("tr").not(".hiddenrow"),
			filter_columnFilters:false,
			zebra : [ "oddRow", "evenRow" ],
			columnSelector_container: $('#' + options.rowheaders + '-table-rows'),
			columnSelector_mediaquery: false,
			columnSelector_saveColumns: false,
			columnSelector_layout: '<div class="columnselectoritem"><input type="checkbox" id="label-sort-{name}"><label for="label-sort-{name}">{name}</label></div>',
			pager_selectors: {
				container: $(options.sorttable + "-pager")
			},
			pager_size:pagerSize,
			pager_savePages:false,
			pager_removeRows:false,
			//pager_fixedHeight: true,
			//pager_storageKey:"alarms-sort-table",
			pager_output: "{page:input}/{filteredPages} (" + pagerSize + ")"			
		}
	});

    console.log(searchArray);
	$.tablesorter.setFilters($("#batch-sort"), searchArray, true);
	console.log($.tablesorter.getFilters($("#batch-sort")));

**Comment by Mottie** [comment]
Hmm, well there are a few other issues:

* `output_saveRows` should be set to `tr:not(.hiddenrow)` (string only) - I'll emphasize this requirement in the documentation.
* `pager_output` can also include `{size}`. You can include the `pagerSize` value in the output, but it won't change if the user changes the pager size. I'll include this in the documentation as well.
* If you are using `"hiddenrow"` as the class that hides the filter row, you'll need to:
  * Make sure the css for `.hiddenrow` is set to `display: none`.
  * Set the `filter_filteredRow` widget option to `"hiddenrow"`

```js
widgetOptions : {
	output_delivery: "download",
	output_saveRows: "tr:not(.hiddenrow)",
	filter_columnFilters: false,
	filter_filteredRow: "hiddenrow",
	// ...
	pager_output: "{page:input}/{filteredPages} ({size})"
}
```

**Comment by doveyg** [comment]
Thank you for clarification on some points and fixing up some of my code.

`.hiiddenrow` is the child row, I am trying to remove those rows from an `output` file. The code was (is?) using the old 2.18 code. This feature wasn't available in that version, as I understand, it was just there for placeholder purposes.

I was unaware that the filtered row class was now explicit and has to be defined in CSS. Thank you.

I have come across another issue though, before the update I was filtering using RegEx.

    [8: "/^Done$|^Aborted$/i"] //Fails - because of the | (or operator)?
    [8: "/^Run$/i"] //Works

I'm assuming there have been some changes to the way the RegEx works in this version too?

**Comment by Mottie** [comment]
Oh ok...

I thought you were expecting the filter row class to be "hiddenrow". All provided themes set the default class to display none ([ref: blue theme](https://github.com/Mottie/tablesorter/blob/master/css/theme.blue.css#L219-L222)). If you are using a provided theme, then no need to modify that option.

So the problem with the regular expression is because of the modified OR filter type. The OR and AND types were updated to allow multiple search types in the filter, e.g. `<10 OR >100`. It is splitting the regex instead of keeping it whole. I'll have that issue fixed here shortly.

And you are correct, I should not have included the "tr" in the `output_saveRows` option. Internally, I set it up to only allow the selector to start with a `.`, `#`, `:` or `[`. I'll update my answer above and add clarification to the documentation.
